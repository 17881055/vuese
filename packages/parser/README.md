# `@vuese/parser`

> TODO: description

## Usage

```
const parser = require('@vuese/parser');

// TODO: DEMONSTRATE API
```
