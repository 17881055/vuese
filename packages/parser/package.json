{
  "name": "@vuese/parser",
  "version": "2.0.0-beta.1",
  "description": "Vue file parser for automatic document generation",
  "keywords": [
    "parser",
    "vue",
    "document"
  ],
  "author": "HcySunYang <HcySunYang@outlook.com>",
  "homepage": "https://github.com/vuese/vuese/tree/master/packages/parser#readme",
  "license": "MIT",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "lib",
    "dist"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vuese/vuese.git"
  },
  "bugs": {
    "url": "https://github.com/vuese/vuese/issues"
  },
  "buildOptions": {
    "extractDts": true
  },
  "dependencies": {
    "@babel/generator": "^7.0.0",
    "@babel/parser": "^7.1.0",
    "@babel/traverse": "^7.1.0",
    "@babel/types": "^7.1.3",
    "@vuese/utils": "^2.0.0",
    "vue-template-compiler": "^2.5.21"
  }
}
