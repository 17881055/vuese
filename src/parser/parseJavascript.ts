import traverse from '@babel/traverse'
import generate from '@babel/generator'
import * as bt from '@babel/types'
import { getComments, CommentResult } from './jscomments'
import {
  PropsResult,
  PropType,
  ParserOptions,
  EventResult,
  MethodResult
} from './index'
import { getValueFromGenerate, isVueOption, runFunction } from '../helpers'
import { isArray } from 'util'

const mainTraveres = {
  ObjectProperty(path: any) {
    const { onProp, onMethod, onName } = (this as any).options
    // Processing name
    if (isVueOption(path, 'name')) {
      if (onName) onName(path.node.value.value)
    }

    // Processing props
    if (isVueOption(path, 'props')) {
      const valuePath = path.get('value')
      let res: PropsResult[] = []
      if (bt.isArrayExpression(valuePath.node)) {
        // An array of strings
        const propsValue: [] = getValueFromGenerate(valuePath.node)
        res = normalizeProps(propsValue)
      } else if (bt.isObjectExpression(valuePath.node)) {
        // An object
        valuePath.traverse({
          ObjectProperty(propPath: any) {
            // Guarantee that this is the prop definition
            if (propPath.parentPath === valuePath) {
              const name = propPath.node.key.name
              const vPath = propPath.get('value')
              const result: PropsResult = {
                name,
                type: null,
                describe: getComments(propPath.node).default
              }

              if (isAllowPropsType(vPath.node)) {
                result.type = getTypeByPath(vPath)
              } else if (bt.isObjectExpression(vPath.node)) {
                if (!vPath.node.properties.length) return

                const allPropNodes = vPath.get('properties.0').container
                const typeNode = allPropNodes.filter((node: any, i: number) => {
                  if (node.key.name === 'type') {
                    node.$$selfPath = vPath.get(`properties.${i}`)
                    return true
                  }
                  return false
                })
                const otherNodes = allPropNodes.filter(
                  (node: any, i: number) => {
                    if (node.key.name !== 'type') {
                      node.$$selfPath = vPath.get(`properties.${i}`)
                      return true
                    }
                    return false
                  }
                )

                // Prioritize `type` before processing `default`.
                // Because the difference in `type` will affect the way `default` is handled.
                if (typeNode.length > 0) {
                  result.type = getTypeByPath(
                    typeNode[0].$$selfPath.get('value')
                  )
                  // Get descriptions of the type
                  const typeDesc: string[] = getComments(typeNode[0]).default
                  if (typeDesc.length > 0) {
                    result.typeDesc = typeDesc
                  }
                }

                otherNodes.forEach((node: any) => {
                  const n = node.key.name
                  if (n === 'default') {
                    if (
                      !hasFunctionTypeDef(result.type) &&
                      bt.isFunction(node.value)
                    ) {
                      result.default = runFunction(node.value)
                    } else {
                      if (bt.isObjectMethod(node)) {
                        result.default = generate(node).code
                      } else {
                        result.default = generate(node.value).code
                      }
                    }

                    // Get descriptions of the default value
                    const defaultDesc: string[] = getComments(node).default
                    if (defaultDesc.length > 0) {
                      result.defaultDesc = defaultDesc
                    }
                  } else if (n === 'required') {
                    if (bt.isBooleanLiteral(node.value)) {
                      result.required = node.value.value
                    }
                  } else if (n === 'validator') {
                    if (bt.isObjectMethod(node)) {
                      result.validator = generate(node).code
                    } else {
                      result.validator = generate(node.value).code
                    }

                    // Get descriptions of the validator
                    const validatorDesc: string[] = getComments(node).default
                    if (validatorDesc.length > 0) {
                      result.validatorDesc = validatorDesc
                    }
                  }
                })
              }
              res.push(result)
            }
          }
        })
      }
      if (onProp) onProp(res)
    }

    // Processing methods
    if (isVueOption(path, 'methods')) {
      const properties = path.node.value.properties
      properties.forEach((node: any) => {
        const commentsRes: CommentResult = getComments(node)
        // Collect only methods that have @vuese annotations
        if (commentsRes.vuese) {
          const result: MethodResult = {
            name: node.key.name,
            describe: commentsRes.default,
            argumentsDesc: commentsRes.arg
          }
          if (onMethod) onMethod(result)
        }
      })
    }
  },
  CallExpression(path: any) {
    const node = path.node
    // this.$emit()
    if (
      bt.isMemberExpression(node.callee) &&
      bt.isThisExpression(node.callee.object) &&
      bt.isIdentifier(node.callee.property) &&
      node.callee.property.name === '$emit' &&
      bt.isExpressionStatement(path.parentPath.node)
    ) {
      const { onEvent } = (this as any).options
      const args = node.arguments
      const result: EventResult = {
        name: ''
      }
      if (args.length && bt.isStringLiteral(args[0])) {
        result.name = args[0].value
      }

      if (!result.name || (this as any).seenEvent.has(result.name)) return
      ;(this as any).seenEvent.add(result.name)

      const allComments: CommentResult = getComments(path.parentPath.node)
      result.describe = allComments.default
      result.argumentsDesc = allComments.arg

      if (onEvent) onEvent(result)
    }
  }
}

function normalizeProps(props: string[]): PropsResult[] {
  return props.map(prop => ({
    type: null,
    name: prop
  }))
}

function getTypeByPath(typePath: any): PropType {
  const typeNode = typePath.node
  if (bt.isIdentifier(typeNode)) return typeNode.name
  if (bt.isArrayExpression(typeNode)) {
    if (!typeNode.elements.length) return null
    return typePath
      .get('elements.0')
      .container.filter((node: any) => bt.isIdentifier(node))
      .map((node: any) => node.name)
  }
  return null
}

// The `type` of a prop should be an array of constructors or constructors
// eg. String or [String, Number]
function isAllowPropsType(typeNode: any): boolean {
  return bt.isIdentifier(typeNode) || bt.isArrayExpression(typeNode)
}

function hasFunctionTypeDef(type: PropType): boolean {
  if (typeof type === 'string') {
    return type.toLowerCase() === 'function'
  } else if (isArray(type)) {
    return type.map(a => a.toLowerCase()).some(b => b === 'function')
  }
  return false
}

export default function(ast: any, options: ParserOptions = {}) {
  const seenEvent = new Set()
  traverse(ast, {
    ExportDefaultDeclaration(path: any) {
      path.traverse(mainTraveres, {
        options,
        seenEvent
      })
    }
  })
}
