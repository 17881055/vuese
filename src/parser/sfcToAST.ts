import { parse as vueSFCParse } from '@vue/component-compiler-utils'
import * as vueTemplateCompiler from 'vue-template-compiler'
import { parse as babelParse, ParserPlugin } from '@babel/parser'
import * as bt from '@babel/types'

export interface AstResult {
  sourceType?: string
  jsAst?: bt.File
  templateAst?: object
}

export default function(
  source: string,
  babelParserPlugins?: ParserPlugin[]
): AstResult {
  const sfc = vueSFCParse({
    source: source,
    compiler: vueTemplateCompiler,
    needMap: false
  })
  const res: AstResult = {}
  if (sfc.script && sfc.script.content) {
    res.sourceType = sfc.script.lang || 'js'
    res.jsAst = babelParse(sfc.script.content, {
      sourceType: 'module',
      plugins: babelParserPlugins || [
        'objectRestSpread',
        'dynamicImport',
        'decorators-legacy',
        'typescript'
      ]
    })
  }
  if (sfc.template && sfc.template.content) {
    res.templateAst = vueTemplateCompiler.compile(sfc.template.content, {
      comments: true
    }).ast
  }
  return res
}
