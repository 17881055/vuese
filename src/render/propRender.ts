import { PropsResult } from '../parser'
import { renderTabelHeader, renderTabelRow } from './tabelRender'
import config from '../config'

export default function(propsRes: PropsResult[]) {
  const propConfig = config.render.props
  let code = renderTabelHeader(propConfig)
  const row: string[] = []
  propsRes.forEach((prop: PropsResult) => {
    for (let i = 0; i < propConfig.length; i++) {
      if (propConfig[i] === 'Name') {
        row.push(prop.name)
      } else if (propConfig[i] === 'Description') {
        let desc: string[] = []
        if (prop.describe) {
          desc = prop.describe
          if (prop.validatorDesc) {
            desc = prop.describe.concat(prop.validatorDesc)
          }
        }
        row.push(desc.join(''))
      } else if (propConfig[i] === 'Type') {
        if (prop.typeDesc) {
          row.push(prop.typeDesc.join(''))
        } else if (!prop.type) {
          row.push('â€”')
        } else if (typeof prop.type === 'string') {
          row.push(`\`${prop.type}\``)
        } else if (Array.isArray(prop.type)) {
          row.push(
            prop.type
              .map(t => `\`${t}\` | `)
              .join('')
              .slice(0, -3)
          )
        } else {
          row.push('-')
        }
      } else if (propConfig[i] === 'Required') {
        if (typeof prop.required === 'undefined') {
          row.push('`false`')
        } else if (typeof prop.required === 'boolean') {
          row.push(`\`${String(prop.required)}\``)
        } else {
          row.push('-')
        }
      } else if (propConfig[i] === 'Default') {
        if (prop.defaultDesc) {
          row.push(prop.defaultDesc.join(''))
        } else if (prop.default) {
          row.push(prop.default)
        } else {
          row.push('-')
        }
      }
    }
  })

  code += renderTabelRow(row)
  return code
}
