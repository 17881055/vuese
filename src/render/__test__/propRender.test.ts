import propRender from '../propRender'
import { PropsResult } from '../../parser'

test('Proper rendering of props', () => {
  const propsRes: PropsResult[] = [
    {
      name: 'propName',
      type: 'String',
      typeDesc: ['`TOP` | `BOTTOM`'],
      required: true,
      defaultDesc: ['`TOP`'],
      validatorDesc: [''],
      describe: ['Direction of the arrow']
    }
  ]
  const str = propRender(propsRes)
  expect(str).toMatchSnapshot()
})
