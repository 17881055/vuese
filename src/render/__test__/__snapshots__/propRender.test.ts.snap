// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Proper rendering of props 1`] = `
"|Name|Description|Type|Required|Default|
=========================================
|propName|Direction of the arrow|\`TOP\` | \`BOTTOM\`|\`true\`|\`TOP\`|
"
`;
